render/render.js

```diff
@@ -543,46 +546,22 @@ module.exports = function($window) {
 		return nextSibling
 	}
 
-	// This covers a really specific edge case:
-	// - Parent node is keyed and contains child
-	// - Child is removed, returns unresolved promise in `onbeforeremove`
-	// - Parent node is moved in keyed diff
-	// - Remaining children still need moved appropriately
-	//
-	// Ideally, I'd track removed nodes as well, but that introduces a lot more
-	// complexity and I'm not exactly interested in doing that.
-	function moveNodes(parent, vnode, nextSibling) {
-		var frag = $doc.createDocumentFragment()
-		moveChildToFrag(parent, frag, vnode)
-		insertNode(parent, frag, nextSibling)
-	}
-	function moveChildToFrag(parent, frag, vnode) {
-		// Dodge the recursion overhead in a few of the most common cases.
-		while (vnode.dom != null && vnode.dom.parentNode === parent) {
-			if (typeof vnode.tag !== "string") {
-				vnode = vnode.instance
-				if (vnode != null) continue
-			} else if (vnode.tag === "<") {
-				for (var i = 0; i < vnode.instance.length; i++) {
-					frag.appendChild(vnode.instance[i])
-				}
-			} else if (vnode.tag !== "[") {
-				// Don't recurse for text nodes *or* elements, just fragments
-				frag.appendChild(vnode.dom)
-			} else if (vnode.children.length === 1) {
-				vnode = vnode.children[0]
-				if (vnode != null) continue
+	// This handles fragments with zombie children (removed from vdom, but persisted in DOM through onbeforeremove)
+	function moveNode(parent, vnode, nextSibling) {
+		if (vnode.dom != null) {
+			var target
+			if (vnode.domSize == null) {
+				// don't allocate for the common case
+				target = vnode.dom
 			} else {
-				for (var i = 0; i < vnode.children.length; i++) {
-					var child = vnode.children[i]
-					if (child != null) moveChildToFrag(parent, frag, child)
-				}
+				target = $doc.createDocumentFragment()
+				for (var dom of domFor(vnode)) target.appendChild(dom)
 			}
-			break
+			insertDOM(parent, target, nextSibling)
 		}
 	}
 
-	function insertNode(parent, dom, nextSibling) {
+	function insertDOM(parent, dom, nextSibling) {
```

Comment:
Can this name change be reverted to align with everything else? Would like to preserve history, plus it doesn't feel necessary.

```suggestion
	function insertNode(parent, dom, nextSibling) {
```

Reply:
We use `xNode` both for functions that deal with vnode generically and for this which is actually a DOM helper.

My naming as is is not consistent either (`removeDOM` and `moveNode` have the same signature now, and similar behavior).

I'd go for DOM for all three, as it clarifies intent (they only deal with DOM concerns), even if two of them take vnode parameters.

Reply:
I *almost* suggested that, just was biased in favor of reducing the diff.

Reply:
I think using DOM in all three cases clarifies intent, and is better than `reallyRemoveFinalDefinitiveNoReallyThisTimeItTheLastOneHAGotchaThereIsMore` (dramatic reenactment :-p).

I moved back `currentDOM` too. It belongs to the final function. It could make sense to scope it as such once we move to ESNext:

```js
//...
module.exports = function($window) {
  // ... 900 lines later
  {
    let currentDOM
    return function(dom, vnodes, redraw) {
      //...
    }
  }
}